rustls-symcrypt/src/verify.rs

```diff
@@ -0,0 +1,383 @@
+use symcrypt::hash::{sha256, sha384, sha512, HashAlgorithm};
+use symcrypt::rsa::{RsaKey, RsaKeyUsage};
+use webpki::alg_id::{self};
+
+/// Rsa signatures from the wire will come in the following ASN1 format:
+/// RSAPublicKey ::= SEQUENCE {
+///     modulus           INTEGER,  -- n
+///     publicExponent    INTEGER   -- e
+/// }
+fn extract_rsa_public_key(pub_key: &[u8]) -> Result<(Vec<u8>, Vec<u8>), InvalidSignature> {
+    let key = AsnRsaPublicKey::try_from(pub_key).map_err(|_| InvalidSignature)?;
+
+    let modulus = key.modulus.as_bytes().to_vec();
+    let exponent = key.public_exponent.as_bytes().to_vec();
+
+    Ok((modulus, exponent))
+}
+
+// NistP256, NistP384, and NistP521 have a prepended Legacy byte that needs to be removed
+fn extract_ecc_public_key(pub_key: &[u8], curve_type: CurveType) -> Vec<u8> {
+    if curve_type == CurveType::Curve25519 {
+        // x25519 does not have the legacy byte prepended
+        return pub_key[..].to_vec();
+    }
+    // remove prepended 0x04 legacy byte
+
+
+    // remove only if 0x04 is there 
+    pub_key[1..].to_vec()
+}
+
+/// Ecc signatures from the wire will come in the following ASN1 format:
+/// ECDSASignature ::= SEQUENCE {
+///     r INTEGER,
+///     s INTEGER
+/// }
+/// SymCrypt expects a concatenated r+s with leading padding and leading 0's for both r and s to be removed
+fn extract_ecc_signature(signature: &[u8], curve: CurveType) -> Result<Vec<u8>, InvalidSignature> {
+    // We use pkcs1::RsaPublicKey because the underlying ASN1 format between an RSA public key and
+    // an ECC signgature is the same.
+    let signature = AsnRsaPublicKey::try_from(signature).map_err(|_| InvalidSignature)?;
+    
+    let component_length = curve.get_size();
+    
+    // Leading 0's are stripped when using as_bytes()
+    // https://docs.rs/pkcs1/0.7.5/pkcs1/struct.UintRef.html
+    let r = signature.modulus.as_bytes(); // cast name from `modulus` to `r`
+    let s = signature.public_exponent.as_bytes(); // cast name from `public_exponent` to `s`
+
+    // SymCrypt takes in a concatenated r+s. with the individual r and s components having a size of curve / 2. 
+    // If for example the curve is P256, the r and s components must individually 16 bytes long.
+    // as_bytes() removes all leading 0s, which covers the case if r or s is 17 bytes long for example.
+    // as_bytes() may remove ALL 0's even if one was randomly generated as part of the signature. 
+    // So after we remove the 0's we prepend 0's that were removed until the lengths are 16.
+    // Majority of the time this will not happen but this is to cover corner cases where it does.
+
+    // Ensure r and s are the correct length (pad with leading zeros if necessary).
+    let mut r_padded = Vec::with_capacity(component_length);
+    let mut s_padded = Vec::with_capacity(component_length);
+
+    // Prepend zeros if r is smaller than component_length.
+    if r.len() < component_length {
```

Comment:
>r.len() < component_length [](http://example.com/codeflow?start=7&length=26)

Isn't it an error if r.len() > component_length.

Same for s.len()

Reply:
Fixed, i believe that this does not need to be addressed though. It would just result in an an error for signature verification in the parent